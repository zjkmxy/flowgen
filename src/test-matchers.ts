import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
