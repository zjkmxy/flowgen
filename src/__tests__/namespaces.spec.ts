import { compiler, beautify } from "..";
import "../test-matchers";

describe("should handle merging with other types", () => {
  describe("function", () => {
    test("interface", async () => {
      const ts = `
declare function test(foo: number): string;
namespace test {
  export interface Foo {
    bar: number
  }
}
`;
      const result = compiler.compileDefinitionString(ts);
      expect(await beautify(result)).toMatchSnapshot();
      expect(result).toBeValidFlowTypeDeclarations();
    });

    test("type", async () => {
      const ts = `
declare function test(foo: number): string;
namespace test {
  export type Foo = {
    bar: number
  }
}
`;
      const result = compiler.compileDefinitionString(ts);
      expect(await beautify(result)).toMatchSnapshot();
      expect(result).toBeValidFlowTypeDeclarations();
    });

    test("const", async () => {
      const ts = `
declare function test(foo: number): string;
namespace test {
  export const ok: number
}
`;
      const result = compiler.compileDefinitionString(ts);
      expect(await beautify(result)).toMatchSnapshot();
      expect(result).toBeValidFlowTypeDeclarations();
    });
  });

  test("class", async () => {
    const ts = `
declare class Album {
  label: Album.AlbumLabel;
}
namespace Album {
  export declare class AlbumLabel { }
}
`;
    const result = compiler.compileDefinitionString(ts);
    expect(await beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations();
  });

  test("enum", async () => {
    const ts = `
// TODO: implement enum merging
enum Color {
  Red = 1,
  Green = 2,
  Blue = 4
}
namespace Color {
  export declare function mixColor(colorName: string): number;
}
`;
    const result = compiler.compileDefinitionString(ts);
    expect(await beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations(); // TODO: prop-missing
  });
});

it("should handle namespaces", async () => {
  const ts = `
namespace test {
  export const ok: number
}
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle namespace merging", async () => {
  const ts = `
namespace test {
  export const ok: number
}
namespace test {
  export const error: string
}
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle namespace function merging", async () => {
  const ts = `
namespace test {
  declare function test(err: number): void
}
namespace test {
  declare function test(response: string): string
}
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle exported interfaces and types", async () => {
  const ts = `
namespace Example {
  export interface StoreModel<S> {}
}
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle nested namespaces", async () => {
  const ts = `
import * as external from "external";

declare namespace E0 {
  type A = external.type;
  namespace U1 {
    declare interface S3 {
      a: string;
    }
  }
  namespace U1 {
    declare var e2: number;
    enum E2 {
      E = 1,
    }
    declare interface S3 {
      b: string;
    }
    namespace D1 {
      namespace S2 {
        declare interface S3 {
          b: string;
        }
        declare var n3: symbol;
        class N3 {}
      }
    }
    namespace DD1 {
      namespace S2 {
        declare interface S3 {
          e: number;
        }
      }
    }
  }
  namespace S1 {
    declare var m3: string;
  }
  declare var s1: string;
}
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations(); // cannot-resolve-module
});

describe("should handle nested namespace merging", () => {
  describe("function", () => {
    test("interface", async () => {
      const ts = `
namespace ns {
  declare function test(foo: number): string;
  namespace test {
    export interface Foo {
      bar: number
    }
  }
}
`;
      const result = compiler.compileDefinitionString(ts);
      expect(await beautify(result)).toMatchSnapshot();
    });

    test("type", async () => {
      const ts = `
namespace ns {
  declare function test(foo: number): string;
  namespace test {
    export type Foo = {
      bar: number
    }
  }
}
`;
      const result = compiler.compileDefinitionString(ts);
      expect(await beautify(result)).toMatchSnapshot();
    });

    test("const", async () => {
      const ts = `
namespace ns {
  declare function test(foo: number): string;
  namespace test {
    export const ok: number
  }
}
`;
      const result = compiler.compileDefinitionString(ts);
      expect(await beautify(result)).toMatchSnapshot();
    });
  });

  test("class", async () => {
    const ts = `
namespace ns {
  declare class Album {
    label: ns.Album.AlbumLabel;
  }
  namespace Album {
    export declare class AlbumLabel { }
  }
}
`;
    const result = compiler.compileDefinitionString(ts);
    expect(await beautify(result)).toMatchSnapshot();
  });

  test("enum", async () => {
    const ts = `
namespace ns {
  // TODO: implement enum merging inside namespaces
  enum Color {
    Red = 1,
    Green = 2,
    Blue = 4
  }
  namespace Color {
    export declare function mixColor(colorName: string): number;
  }
}
`;
    const result = compiler.compileDefinitionString(ts);
    expect(await beautify(result)).toMatchSnapshot();
  });
});

test("should handle qualified namespaces", async () => {
  const ts = `
declare namespace A.B {
  interface S<A> {
    readonly d: A;
    b: number;
}
  declare class D<S> {}
}

declare namespace A.B.C {
  declare class N<A> extends D<A> implements S<A> {
    a: string;
  }
}`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations(); // TODO: type-as-value
});

test("should handle global augmentation", async () => {
  const ts = `
declare global {
  interface Array<T> {}
}`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

test("should handle import equals declaration", async () => {
  const ts = `
namespace A {
  export type B = string;
}
import hello = A.B;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
