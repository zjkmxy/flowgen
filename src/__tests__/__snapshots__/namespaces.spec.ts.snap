// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle exported interfaces and types 1`] = `
"declare namespace Example {
  declare export interface StoreModel<S> {}
}
"
`;

exports[`should handle global augmentation 1`] = `
"declare module "global" {
  declare interface Array<T> {}
}
"
`;

exports[`should handle import equals declaration 1`] = `
"declare namespace A {
  declare export type B = string;
}

declare type hello = A.B;
"
`;

exports[`should handle merging with other types class 1`] = `
"declare class Album {
  label: Album.AlbumLabel;
  static AlbumLabel: typeof Album.AlbumLabel;
}

declare export class AlbumLabel {}
"
`;

exports[`should handle merging with other types enum 1`] = `
"declare enum Color {
  Red = 1, // 1
  Green = 2, // 2
  Blue = 4, // 4
}

declare namespace Color {
  declare export function mixColor(colorName: string): number;
}
"
`;

exports[`should handle merging with other types function const 1`] = `
"declare namespace test {
  declare export var ok: number;
}
"
`;

exports[`should handle merging with other types function interface 1`] = `
"declare namespace test {
  declare export interface Foo {
    bar: number;
  }
}
"
`;

exports[`should handle merging with other types function type 1`] = `
"declare namespace test {
  declare export type Foo = {
    bar: number,
  };
}
"
`;

exports[`should handle namespace function merging 1`] = `
"declare namespace test {
  declare function test(err: number): void;

  declare function test(response: string): string;
}
"
`;

exports[`should handle namespace merging 1`] = `
"declare namespace test {
  declare export var test.ok: number;
  declare export var test.error: string;
}
"
`;

exports[`should handle namespaces 1`] = `
"declare namespace test {
  declare export var test.ok: number;
}
"
`;

exports[`should handle nested namespace merging class 1`] = `
"declare namespace ns {
  declare class Album {
    label: ns.Album.AlbumLabel;
    static AlbumLabel: typeof Album.AlbumLabel;
  }

  declare export class AlbumLabel {}
}
"
`;

exports[`should handle nested namespace merging enum 1`] = `
"declare namespace ns {
  declare enum Color {
    Red = 1, // 1
    Green = 2, // 2
    Blue = 4, // 4
  }

  declare namespace Color {
    declare export function mixColor(colorName: string): number;
  }
}
"
`;

exports[`should handle nested namespace merging function const 1`] = `
"declare namespace ns {
  declare namespace test {
    declare export var test.ok: number;
  }
}
"
`;

exports[`should handle nested namespace merging function interface 1`] = `
"declare namespace ns {
  declare namespace test {
    declare export interface Foo {
      bar: number;
    }
  }
}
"
`;

exports[`should handle nested namespace merging function type 1`] = `
"declare namespace ns {
  declare namespace test {
    declare export type Foo = {
      bar: number,
    };
  }
}
"
`;

exports[`should handle nested namespaces 1`] = `
"import * as external from "external";

declare namespace E0 {
  declare type A = external.type;

  declare namespace U1 {
    declare interface S3 {
      a: string;
      b: string;
    }
    declare var E0.U1.e2: number;

    declare enum E2 {
      E = 1, // 1
    }

    declare namespace D1 {
      declare namespace S2 {
        declare interface S3 {
          b: string;
        }
        declare var E0.U1.D1.S2.n3: Symbol;
        declare class N3 {}
      }
    }

    declare namespace DD1 {
      declare namespace S2 {
        declare interface S3 {
          e: number;
        }
      }
    }
  }

  declare namespace S1 {
    declare var E0.S1.m3: string;
  }

  declare var E0.s1: string;
}
"
`;

exports[`should handle qualified namespaces 1`] = `
"declare namespace A {
  declare namespace B {
    declare interface S<A> {
      +d: A;
      b: number;
    }
    declare class D<S> {}

    declare namespace C {
      declare class N<A> mixins A.B.D<A> implements A.B.S<A> {
        a: string;
      }
    }
  }
}
"
`;
